pkg/generic/descriptor/http.go

```diff
@@ -32,25 +35,124 @@ const (
 	MIMEApplicationProtobuf = "application/x-protobuf"
 )
 
+var (
+	_ dhttp.RequestGetter  = &HTTPRequest{}
+	_ dhttp.ResponseSetter = &HTTPResponse{}
+)
+
 // HTTPRequest ...
 type HTTPRequest struct {
-	Header      http.Header
-	Query       url.Values
-	Cookies     Cookies
-	Method      string
-	Host        string
-	Path        string
 	Params      *Params // path params
+	Request     *http.Request
 	RawBody     []byte
 	Body        map[string]interface{}
 	GeneralBody interface{} // body of other representation, used with ContentType
 	ContentType MIMEType
+	cookies     map[string]string
+	query       url.Values
+	bodyMap     *ast.Node
+}
+
+// GetHeader implements http.RequestGetter of dynamicgo
+func (req *HTTPRequest) GetHeader(key string) string {
+	return req.Request.Header.Get(key)
+}
+
+// GetCookie implements http.RequestGetter of dynamicgo
+func (req *HTTPRequest) GetCookie(key string) string {
+	if req.cookies == nil {
+		req.cookies = map[string]string{}
+		for _, cookie := range req.Request.Cookies() {
+			req.cookies[cookie.Name] = cookie.Value
+		}
+	}
+	return req.cookies[key]
+}
+
+// GetQuery implements http.RequestGetter of dynamicgo
+func (req *HTTPRequest) GetQuery(key string) string {
+	if req.Request.URL == nil {
+		return ""
+	}
+	if req.query == nil {
+		req.query = req.Request.URL.Query()
+	}
+	return req.query.Get(key)
+}
+
+// GetBody implements http.RequestGetter of dynamicgo
+func (req *HTTPRequest) GetBody() []byte {
+	return req.RawBody
+}
+
+// GetMethod implements http.RequestGetter of dynamicgo
+func (req *HTTPRequest) GetMethod() string {
+	return req.Request.Method
+}
+
+// GetPath implements http.RequestGetter of dynamicgo
+func (req *HTTPRequest) GetPath() string {
+	return req.Request.URL.Path
```

Comment:
Since URL might be nil, why don't you check nil here?

Reply:
fixed